File: amm/compiler/interface/src/main/java/ammonite/compiler/iface/CompilerBuilder.java
package ammonite.compiler.iface;

import java.net.URL;
import java.util.function.Consumer;
import java.util.Map;

public abstract class CompilerBuilder {

  public abstract Compiler create(
    URL[] initialClassPath,
    URL[] classPath,
    Map.Entry<String, byte[]>[] dynamicClassPath,
    ClassLoader evalClassLoader,
    ClassLoader pluginClassLoader,
    Consumer<Message> reporter,
    String[] settings,
    String[][] classPathWhiteList,
    boolean lineNumberModifier
  );

  public abstract String userScalaVersion();

  public static final class Message {
    private final String severity;
    private final int start;
    private final int end;
    private final String message;

    public String severity() {
      return severity;
    }
    public int start() {
      return start;
    }
    public int end() {
      return end;
    }
    public String message() {
      return message;
    }

    public Message(
      String severity,
      int start,
      int end,
      String message
    ) {
      this.severity = severity;
      this.start = start;
      this.end = end;
      this.message = message;
    }
  }

}

File: amm/compiler/interface/src/main/scala/ammonite/compiler/iface/CompilerBuilder.scala
package ammonite.compiler.iface

import java.net.URL
import java.nio.file.Path

import ammonite.util.Frame

abstract class CompilerBuilder {

  def newManager(
    rtCacheDir: Option[Path],
    headFrame: => Frame,
    dependencyCompleter: => Option[String => (Int, Seq[String])],
    whiteList: Set[Seq[String]],
    initialClassLoader: ClassLoader,
    settings: Seq[String]
  ): CompilerLifecycleManager

  def create(
    initialClassPath: Seq[URL],
    classPath: Seq[URL],
    dynamicClassPath: Seq[(String, Array[Byte])],
    evalClassLoader: ClassLoader,
    pluginClassLoader: ClassLoader,
    reporter: Option[CompilerBuilder.Message => Unit],
    settings: Seq[String],
    classPathWhiteList: Set[Seq[String]],
    lineNumberModifier: Boolean
  ): Compiler

  def scalaVersion: String
}

object CompilerBuilder {

  case class Message(
    severity: String,
    start: Int,
    end: Int,
    message: String
  )

}
